### PR TITLE
Generalized iterated unification

### PR DESCRIPTION
**Pull request by virgil-serbanuta**
Not fully ready for review, but it kind of works.

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

**Comment by virgil-serbanuta** [comment]
@traiansf I'm unsure about the variable renaming stuff - I could either check that a variable has the new variable format and then I would only change the suffix, or I could roll it back - what do you think? This PR's implementation is not that good - it can make longer and longer variable names.

File: src/main/haskell/kore/src/Kore/Variables/Fresh.hs [hunk 1]
```diff
@@ -59,19 +59,40 @@ a non-id symbol @_@ to avoid clashing with user-defined ids.
 freshVariablePrefix :: Text
 freshVariablePrefix = "var_"
 
+variableSeparator :: Text
+variableSeparator = "_"
+
 instance FreshVariable Variable where
```
Comment (traiansf) [inline src/main/haskell/kore/src/Kore/Variables/Fresh.hs]: Would you mind documenting the syntax of fresh variables?  It seems to get more complex now.

File: src/main/haskell/kore/test/Test/Kore/Step/BaseStep.hs [hunk 2]
```diff
@@ -475,17 +475,12 @@ test_baseStep =
...
-                , predicate = give mockSymbolOrAliasSorts $ makeEqualsPredicate
-                    (asPureMetaPattern (b1 PatternSort))
-                    (asPureMetaPattern (metaH (b1 PatternSort)))
-                , substitution =
-                    [   ( asMetaVariable (a1 PatternSort)
-                        , asPureMetaPattern (metaH (b1 PatternSort))
-                        )
-                    ]
-                }
+            (Left
```
Comment (traiansf) [inline src/main/haskell/kore/test/Test/Kore/Step/BaseStep.hs]: Add a TODO with the previous result.

File: src/main/haskell/kore/test/Test/Kore/Step/Function/Integration.hs [hunk 3]
```diff
@@ -233,9 +239,49 @@ test_functionIntegration = give mockSymbolOrAliasSorts
...
+        -- 2. Transforming the 'and' in an equals predicate,
+        --    as it would happen for functions.
+        (assertEqualWithExplanation ""
+            Predicated
+                { term = Mock.a
+                , predicate =
+                    makeCeilPredicate
+                        (Mock.plain10 Mock.cf)
+                , substitution =
+                    [ (Mock.var_x_1, Mock.cf), (Mock.var_y_1, Mock.b) ]
```
Comment (traiansf) [inline src/main/haskell/kore/test/Test/Kore/Step/Function/Integration.hs]: Why are the variables renamed here?

File: src/main/haskell/kore/src/Kore/Builtin/Map.hs [hunk 4]
```diff
@@ -569,41 +571,62 @@ unify
...
+            unified = give symbolOrAliasSorts $ do  -- MonadCounter
+                (intersectMapWithProofs :: Map k (expanded, proof)) <-
+                    collectProvenExpandedPatterns unifiedIntersect
+                let
+                    intersectMap :: Map k expanded
+                    intersectMap =
+                        discardProofs intersectMapWithProofs
+                    propagated :: Predicated level variable (Map k p)
+                    propagated = propagatePredicates intersectMap
+                    q :: expanded
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply src/main/haskell/kore/src/Kore/Builtin/Map.hs]: Done.

File: src/main/haskell/kore/src/Kore/Step/BaseStep.hs [hunk 5]
```diff
@@ -311,40 +317,21 @@ stepWithAxiom'
             , substitution = normalizedSubstitution
             }
         , _  -- TODO: Use this proof
-        ) <- stepperVariableToVariableForError
-            existingVars
-            $ mapExceptT (fmap unificationOrSubstitutionToStepError)
-            $ mergeAndNormalizeSubstitutions tools rawSubstitution startSubstitution
-    let
-        -- Merge all conditions collected so far
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply src/main/haskell/kore/src/Kore/Step/BaseStep.hs]: I merged and reworded the two comments.

File: src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs [hunk 6]
```diff
@@ -97,17 +100,22 @@ matchAsUnification
...
+        (UnificationOrSubstitutionError level variable)
         m
         ( PredicateSubstitution level variable
         , UnificationProof level variable
         )
-matchAsUnification tools first second =
-    noteT UnsupportedPatterns
-    $ (,) <$> match tools Map.empty first second <*> return EmptyUnificationProof
+matchAsUnification tools substitutionSimplifier first second = do
+    result <- runMaybeT matchResult
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs]: It's not possible, match now returns an ErrorT under a MaybeT.

File: src/main/haskell/kore/src/Kore/Step/Simplification/PredicateSubstitution.hs [hunk 7]
```diff
@@ -116,35 +116,42 @@ simplify
...
+            if null simplifiedSubstitution
+                then return
+                    ( PredicateSubstitution
+                        { predicate = simplifiedPredicate
+                        , substitution
+                        }
+                    , SimplificationProof
+                    )
+                else do
+                    assertDistinctVariables
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply src/main/haskell/kore/src/Kore/Step/Simplification/PredicateSubstitution.hs]: Added TODO.

File: src/main/haskell/kore/test/Test/Kore/Step/Function/Integration.hs (hunk 3 above)
Comment (virgil-serbanuta) [reply src/main/haskell/kore/test/Test/Kore/Step/Function/Integration.hs]: The test framework renames them below.

**Comment by virgil-serbanuta** [comment]
Jenkins: test this please.

**Pull request merged**